### PR TITLE
Webpack 4 ✨

### PR DESCRIPTION
**Pull request by brandonc**
I had some time last night to upgrade to Webpack 4. It's slightly faster, but doesn't offer any big advantages that I know of right now. If we migrate to babel 7, we can drop ts-loader since babel 7 can transpile typescript now. That could potentially knock 20s off our build time

## Testing

`npm start` or `npm run build-assets`

## Trade-offs

I don't know of any risks.

## Dependencies

This branch is based on #3307

### CODE REVIEW

**Comment by brandonc** [comment]
TIL you can't reopen a PR after force pushing it (#3310)

**Comment by nLight** [comment]
I'd like to hold merge this until we branch `release/1.12`

File: webpack/webpack.production.js [hunk 1]
```diff
@@ -23,30 +24,8 @@ module.exports = merge(common, {
   plugins: [
     new ModuleConcatenationPlugin(),
     new DefinePlugin({
-      "process.env.NODE_ENV": JSON.stringify("production"),
       "process.env.version": JSON.stringify(packageInfo.version)
     }),
-    new UglifyJsPlugin({
```
Comment (nLight) [inline webpack/webpack.production.js]: Anything else to replace uglify?

File: webpack/webpack.production.js [hunk 2]
```diff
@@ -15,6 +15,7 @@ delete dependencies["canvas-ui"];
 delete dependencies["cnvs"];
 
 module.exports = merge(common, {
+  mode: "production",
```
Comment (nLight) [inline webpack/webpack.production.js]: will this replace `"process.env.NODE_ENV"`?

File: webpack/webpack.config.js [hunk 3]
```diff
@@ -65,6 +65,12 @@ module.exports = {
   ],
   module: {
     rules: [
+      {
+        type: "javascript/auto",
```
Comment (nLight) [inline webpack/webpack.config.js]: Would that transpile all the modules? Do we need this or it's more like a nice to have?

File: webpack/webpack.production.js (hunk 2 above)
Comment (brandonc) [reply webpack/webpack.production.js]: Yes; webpack sets this and other behaviors like uglify automatically in this mode

File: webpack/webpack.config.js (hunk 3 above)
Comment (brandonc) [reply webpack/webpack.config.js]: Not a long term solution, but a workaround owing the to the fact that webpack 4 prefers mjs files to js files and react-apollo fails to account for this. https://github.com/apollographql/react-apollo/issues/1737

File: webpack/webpack.production.js (hunk 2 above)
Comment (brandonc) [reply webpack/webpack.production.js]: I didn't intend for this to be a surprise but left the background in the commit messages . I'll be sure to add these to the PR description in the future https://webpack.js.org/concepts/mode/

Comment (nLight) [reply webpack/webpack.production.js]: Oh, my bad. Don't have a strong habit of reading commit messages during a review

**Comment by nLight** [comment]
Conflicts!

**Comment by brandonc** [comment]
Ill merge #3307 and then rebase on master

**Comment by nLight** [comment]
I'd say the longer we wait the closer to a release we will be merging this :)

**Comment by GeorgiSTodorov** [comment]
Jira: https://jira.mesosphere.com/browse/DCOS_OSS-2358

**Pull request merged**